File: backend/app/models/task.py
import uuid
from sqlalchemy import Column, String, DateTime, ForeignKey, Text
from sqlalchemy.sql import func
from ..database import Base

def generate_uuid():
    return str(uuid.uuid4())

class Task(Base):
    __tablename__ = "tasks"

    id = Column(String, primary_key=True, index=True, default=generate_uuid)
    title = Column(String, nullable=False)
    description = Column(Text, default="")
    status = Column(String, default="todo") # todo, in_progress, review, done
    priority = Column(String, default="medium") # low, medium, high, critical
    due_date = Column(DateTime(timezone=True), nullable=True)
    
    project_id = Column(String, ForeignKey("projects.id"), nullable=False, index=True)
    assignee_id = Column(String, ForeignKey("users.id"), nullable=True, index=True)
    
    # Similar to Project, we might have attachments or tags, stored as strings
    attachment_urls = Column(String, default="")
    
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())

File: ims-platform/backend/src/models/Task.js
'use strict';

const mongoose = require('mongoose');

const taskSchema = new mongoose.Schema(
    {
        title: { type: String, required: [true, 'Task title is required'], trim: true },
        description: { type: String, default: '' },
        status: {
            type: String,
            enum: ['todo', 'in_progress', 'review', 'done'],
            default: 'todo',
        },
        priority: {
            type: String,
            enum: ['low', 'medium', 'high', 'critical'],
            default: 'medium',
        },
        dueDate: { type: Date },

        // ── Relations ──────────────────────────────────────────────────────────────
        projectId: { type: mongoose.Schema.Types.ObjectId, ref: 'Project', required: true, index: true },
        assigneeId: { type: mongoose.Schema.Types.ObjectId, ref: 'User', default: null },
        createdBy: { type: mongoose.Schema.Types.ObjectId, ref: 'User' },

        // ── Attachments ────────────────────────────────────────────────────────────
        attachmentUrls: [{ type: String }],

        // ── Soft delete ────────────────────────────────────────────────────────────
        deletedAt: { type: Date, default: null },
    },
    { timestamps: true }
);

taskSchema.index({ projectId: 1, status: 1 });
taskSchema.index({ assigneeId: 1, status: 1 });
taskSchema.index({ dueDate: 1 });

taskSchema.pre(/^find/, function (next) {
    if (!this._conditions.deletedAt) this.where({ deletedAt: null });
    next();
});

module.exports = mongoose.model('Task', taskSchema);
